pkg/ruler/rulestore/bucket_client.go

```diff
@@ -0,0 +1,285 @@
+	ch := make(chan *rules.RuleGroupDesc)
+
+	// Given we store one file per rule group. With this, we create a pool of workers that will
+	// download all rule groups in parallel. We limit the number of workers to avoid a
+	// particular user having too many rule groups rate limiting us with the object storage.
+	g, gCtx := errgroup.WithContext(ctx)
+	for i := 0; i < b.loadConcurrency; i++ {
+		g.Go(func() error {
+			for gr := range ch {
+				user, namespace, group := gr.GetUser(), gr.GetNamespace(), gr.GetName()
+				if user == "" || namespace == "" || group == "" {
+					return fmt.Errorf("invalid rule group: user=%q, namespace=%q, group=%q", user, namespace, group)
+				}
+
+				gr, err := b.getRuleGroup(gCtx, user, namespace, group, gr) // reuse group pointer from the map.
+				if err != nil {
+					level.Error(b.logger).Log("msg", "failed to get rule group", "namespace", namespace, "name", group, "user", user, "err", err)
+					return err
+				}
+
+				if user != gr.User || namespace != gr.Namespace || group != gr.Name {
+					return fmt.Errorf("mismatch between requested rule group and loaded rule group, requested: user=%q, namespace=%q, group=%q, loaded: user=%q, namespace=%q, group=%q", user, namespace, group, gr.User, gr.Namespace, gr.Name)
+				}
+			}
+
+			return nil
+		})
+	}
+
+outer:
+	for _, gs := range groupsToLoad {
+		for _, g := range gs {
+			if g == nil {
+				continue
+			}
+			select {
+			case <-gCtx.Done():
+				break outer
+			case ch <- g:
+				// ok
+			}
+		}
+	}
+	close(ch)
+
+	return g.Wait()
+}
+
+func (b *BucketRuleStore) GetRuleGroup(ctx context.Context, userID string, namespace string, group string) (*rules.RuleGroupDesc, error) {
+	return b.getRuleGroup(ctx, userID, namespace, group, nil)
+}
+
+func (b *BucketRuleStore) SetRuleGroup(ctx context.Context, userID string, namespace string, group *rules.RuleGroupDesc) error {
+	userBucket := bucket.NewUserBucketClient(userID, b.bucket)
+	data, err := proto.Marshal(group)
+	if err != nil {
+		return err
+	}
+
+	return userBucket.Upload(ctx, generateRuleObjectKey(namespace, group.Name), bytes.NewBuffer(data))
+}
+
+func (b *BucketRuleStore) DeleteRuleGroup(ctx context.Context, userID string, namespace string, group string) error {
+	userBucket := bucket.NewUserBucketClient(userID, b.bucket)
+	err := userBucket.Delete(ctx, generateRuleObjectKey(namespace, group))
+	if b.bucket.IsObjNotFoundErr(err) {
+		return rules.ErrGroupNotFound
+	}
+	return err
+}
+
+func (b *BucketRuleStore) DeleteNamespace(ctx context.Context, userID string, namespace string) error {
+	ruleGroupList, err := b.ListRuleGroupsForUserAndNamespace(ctx, userID, namespace)
+	if err != nil {
+		return err
+	}
+
+	if len(ruleGroupList) == 0 {
+		return rules.ErrGroupNamespaceNotFound
+	}
+
+	for _, rg := range ruleGroupList {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		userBucket := bucket.NewUserBucketClient(userID, b.bucket)
```

Comment:
We should only create `userBucket` once, not for every rule group.

Reply:
👍